src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java

```diff
@@ -208,67 +269,59 @@ public void periodic() {
     }
 
     // Update odometry
-    int deltaCount =
-        Math.min(
-            gyroInputs.connected ? gyroInputs.odometryYawPositions.length : Integer.MAX_VALUE,
-            Arrays.stream(modules)
-                .map((m) -> m.getPositionDeltas().length)
-                .min(Integer::compare)
-                .get());
-    for (int deltaIndex = 0; deltaIndex < deltaCount; deltaIndex++) {
+    double[] sampleTimestamps =
+        modules[0].getOdometryTimestamps(); // All signals are sampled together
+    int sampleCount = sampleTimestamps.length;
+
+    // int deltaCount =
+    //     Math.min(
+    //         gyroInputs.connected ? gyroInputs.odometryYawPositions.length : Integer.MAX_VALUE,
+    //         Arrays.stream(modules)
+    //             .map((m) -> m.getPositionDeltas().length)
+    //             .min(Integer::compare)
+    //             .get());
+    for (int deltaIndex = 0; deltaIndex < sampleCount; deltaIndex++) {
       // Read wheel deltas from each module
       SwerveModulePosition[] wheelDeltas = new SwerveModulePosition[4];
+      SwerveModulePosition[] modulePositions = getModulePositions();
       for (int moduleIndex = 0; moduleIndex < 4; moduleIndex++) {
-        wheelDeltas[moduleIndex] = modules[moduleIndex].getPositionDeltas()[deltaIndex];
+        wheelDeltas[moduleIndex] = modules[moduleIndex].getOdometryPositions()[deltaIndex];
         modulePositions[moduleIndex] =
             new SwerveModulePosition(
                 modulePositions[moduleIndex].distanceMeters
-                    + wheelDeltas[moduleIndex].distanceMeters,
-                modulePositions[moduleIndex].angle.plus(wheelDeltas[moduleIndex].angle));
+                    - lastModulePositions[moduleIndex].distanceMeters,
+                modulePositions[moduleIndex].angle);
+        lastModulePositions[moduleIndex] = modulePositions[moduleIndex];
       }
 
-      // The twist represents the motion of the robot since the last
-      // sample in x, y, and theta based only on the modules, without
-      // the gyro. The gyro is always disconnected in simulation.
-      var twist = kinematics.toTwist2d(wheelDeltas);
       if (gyroInputs.connected) {
-        // If the gyro is connected, replace the theta component of the twist
-        // with the change in angle since the last sample.
-        Rotation2d gyroRotation = gyroInputs.odometryYawPositions[deltaIndex];
-        twist = new Twist2d(twist.dx, twist.dy, gyroRotation.minus(lastGyroRotation).getRadians());
-        lastGyroRotation = gyroRotation;
+        rawGyroRotation = gyroInputs.odometryYawPositions[deltaIndex];
+      } else {
+        // Use the angle delta from the kinematics and module deltas
+        Twist2d twist = kinematics.toTwist2d(wheelDeltas);
+        rawGyroRotation = rawGyroRotation.plus(new Rotation2d(twist.dtheta));
+        pose = pose.exp(twist);
```

Comment:
shouldnt this be called no matter if the gyro is connected or not?

Reply:
i don't think so because it seems like a replacement for the gyro